pkg/kernel/networkservice/connectioncontextkernel/ipcontext/iprule/common.go

```diff
@@ -54,52 +56,73 @@ func create(ctx context.Context, conn *networkservice.Connection, tableIDs *Map)
 		if err != nil {
 			return errors.Wrapf(err, "failed to find link %s", ifName)
 		}
-
-		ps, ok := tableIDs.Load(conn.GetId())
+		connID := conn.GetId()
+		ps, ok := tableIDs.Load(connID)
 		if !ok {
 			if len(conn.Context.IpContext.Policies) == 0 {
 				return nil
 			}
 			ps = make(map[int]*networkservice.PolicyRoute)
-			tableIDs.Store(conn.GetId(), ps)
+			tableIDs.Store(connID, ps)
 		}
 		// Get policies to add and to remove
 		toAdd, toRemove := getPolicyDifferences(ps, conn.Context.IpContext.Policies)
 
 		// Remove no longer existing policies
 		for tableID, policy := range toRemove {
-			if err := delRule(ctx, netlinkHandle, policy, tableID); err != nil {
-				return err
+			if errRule := delRule(ctx, netlinkHandle, policy, tableID); errRule != nil {
+				return errRule
 			}
 			delete(ps, tableID)
-			tableIDs.Store(conn.GetId(), ps)
+			tableIDs.Store(connID, ps)
 		}
 		// Add new policies
+		netNSInode := mechanism.GetNetNSInode()
 		for _, policy := range toAdd {
-			tableID, err := getFreeTableID(ctx, netlinkHandle)
-			if err != nil {
-				return err
-			}
-			// If policy doesn't contain any route - add default
-			if len(policy.Routes) == 0 {
-				policy.Routes = append(policy.Routes, defaultRoute())
-			}
-
-			for _, route := range policy.Routes {
-				if err := routeAdd(ctx, netlinkHandle, l, route, tableID); err != nil {
+			var tableID int
+			var nsrtid *NetnsRTableNextID
+			// get a free table ID until we succeed
+			for {
+				tableID, err = getFreeTableID(ctx, netlinkHandle)
+				if err != nil {
 					return err
 				}
+				nsrtid = NewNetnsRTableNextID(netNSInode, tableID)
+				storedConnID, _ := nsRTableNextIDToConnID.LoadOrStore(*nsrtid, connID)
+				if connID == storedConnID {
+					// No other connection adding policy using this free routing table ID
+					break
+				}
 			}
-			if err := ruleAdd(ctx, netlinkHandle, policy, tableID); err != nil {
+			if err := addPolicy(ctx, netlinkHandle, policy, l, ps, tableIDs, tableID, connID, *nsrtid, nsRTableNextIDToConnID); err != nil {
 				return err
 			}
-			ps[tableID] = policy
-			tableIDs.Store(conn.GetId(), ps)
 		}
 	}
 	return nil
 }
 
+func addPolicy(ctx context.Context, netlinkHandle *netlink.Handle, policy *networkservice.PolicyRoute, l netlink.Link, ps policies, tableIDs *Map, tableID int, connID string, nsrtid NetnsRTableNextID, nsRTableNextIDToConnID *NetnsRTableNextIDToConnMap) error {
+	// If policy doesn't contain any route - add default
```

Comment:
```suggestion
	// release the lock
	nsRTableNextIDToConnID.Delete(nsrtid)
	// If policy doesn't contain any route - add default
```

If the routes / policy routes fails to be added, the lock will never be released (line 122)